feat(storage): Added Thread Safety to AdjacencyList Storage Format

me-hem opened the pull request:
This PR adds thread safety to `PeakStore::AdjacencyList`by introducing coarse-grained synchronization with `std::shared_mutex` (issue #65 ). Write operations such as `impl_addVertex`, `impl_addVertices`,  `impl_addEdge`  and `impl_addEdges` now use `std::unique_lock` for exclusive access, while read operations like `impl_getEdge`, `impl_getNeighbors`, `impl_doesEdgeExist`, and `print_adj_list` use `std::shared_lock` to allow concurrent reads. This ensures data integrity under concurrent workloads, prevents race conditions or memory corruption, and enables safe multi-threaded access to the storage engine.

### Changes made:

-  `AdjacencyList.hpp` -  Modified methods to use appropriate locking for safe concurrent access.
-  `AdjacencyShard.cpp` - Added tests to validate the thread safety implementation.

me-hem committed:
Locks added to AdjacencyList for thread safety.

me-hem committed:
Tests for thread safety added to AdjacencyShard

me-hem committed:
Thread safety implementation tested successfully

SharonIV0x86 commented:
Adding this for future reference, ``addEdges`` needs to be atomic, Bulk operations don't provide atomicity some operations may succeed while others fail if the vertices dont exist. We will need a all-or-nothing primitive, if any of the vertices do not exist.

Possibly a two phase locking + commit strategy. And the same goes for ``addVertices`` as well. But it can cause [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use). Single lock seems the best options first validate and then write, which can cause the lock to be held for a little longer. But i would prefer safety than just letting some other thread modify while switching read lock to write lock, leaving the storage format in a inconsistent state.

me-hem commented:
@SharonIV0x86 Will work on this later, as these APIs need to be optimized as well.

SharonIV0x86 commented:
> @SharonIV0x86 Will work on this later, as these APIs need to be optimized as well.

No issues, this is not related to the PR anyways. Its just added so that i can reference this.

me-hem commented:
@SharonIV0x86, I have resolved the conflicts. Now, you can consider reviewing.

SharonIV0x86 commented:
@me-hem Your locking strategy is fine, consider this PR approved from my side, but i am thinking about many more scenarios of deadlocks and race conditions which can occur if not handled correctly so it will take some time for me to review and document those issues here just for reference and once done i will proceed with the merge. And i might make changes to the tests or the code.

SharonIV0x86 committed:
Update AdjacencyShard.cpp

Added few more test cases for testing of possible deadlocks, reentrant locks and race conditions.

me-hem commented:
@SharonIV0x86 Okay, I will now proceed with the remaining issues.

SharonIV0x86 committed:
Update AdjacencyList.hpp

Added a Thread safety note for deadlocks.

SharonIV0x86 commented:
@me-hem Kindly resolve the conflicts, i am planning to merge all by tomorrow.

me-hem commented:
@SharonIV0x86, I have resolved the conflicts. Now, you can consider merging.

SharonIV0x86 committed:
fix: fixed files using clang-format

SharonIV0x86 committed:
fix: fixed files using clang-format

SharonIV0x86 reviewed (approved):
Approved.

SharonIV0x86 commented:
@me-hem Thank you for contributing your PR has been merged.
